cellmlmanip/model.py

```diff
@@ -619,6 +619,205 @@ def variables(self):
+
+        # 2. create new equation and remove original
+        expression = sympy.Eq(new_deriv_variable, eqn.args[1])
+        self.add_equation(expression)
+        self.remove_equation(eqn)
+
+        return new_deriv_variable
+
+    def _convert_free_variable_deriv(self, eqn, new_variable, cf):
+        """
+        Create relevant variables/equations when converting a free variable.
+        :param eqn: the derivative equation containing free variable
+        :param new_variable: the new variable representing the converted symbol
+        :param cf: conversion factor for unit conversion
+        """
+        derivative_variable = eqn.args[0].args[0]
+        # 1. create a new variable/equation for derivative
+        new_deriv_variable = self._create_new_deriv_variable_and_equation(eqn, derivative_variable)
+
+        # 3. create equation for derivative wrt new variable
+        expression = sympy.Eq(sympy.Derivative(derivative_variable, new_variable), new_deriv_variable / cf)
+        self.add_equation(expression)
+        return {'variable': new_deriv_variable, 'expression': eqn.args[0]}
+
+    def _convert_state_variable_deriv(self, original_variable, new_variable, cf):
+        """
+        Create relevant variables/equations when converting a state variable.
+        :param original_variable: the variable to be converted
+        :param new_variable: the new variable representing the converted symbol
+        :param cf: conversion factor for unit conversion
+        :return: a dictionary containing the 'variable' and 'expression' for new derivative
+        """
+        # 1. find the derivative equation for this variable
+        # and get the free variable (wrt_variable)
+        eqn = None
+        for equation in self.equations:
+            if equation.args[0].is_Derivative:
+                if equation.args[0].args[0] == original_variable:
+                    eqn = equation
+                    break
+
+        # get free variable symbol
+        wrt_variable = eqn.args[0].args[1]
+
+        # 2. create new variable for derivative
+        new_deriv_variable = self._create_new_deriv_variable_and_equation(eqn, original_variable)
+
+        # 3. add a new derivative equation
+        expression = sympy.Eq(sympy.Derivative(new_variable, wrt_variable), new_deriv_variable * cf)
+        self.add_equation(expression)
+        return {'variable': new_deriv_variable, 'expression': eqn.args[0]}
+
+    def _convert_variable_instance(self, original_variable, cf, units):
+        """
+        Internal function to create new variable and an equation for it.
+        :param original_variable: VariableDummy object to be converted
+        :param cf: conversion factor
+        :param units: Unit object for new units
+        :return: the new variable created
+        """
+        # 1. get unique name for new variable
+        new_name = original_variable.name + '_converted'
```

Comment:
This 3-line bit of logic is used at least twice; worth extracting a helper method?